File: gfv_robot_ruby/robot.rb
class Robot
  VALID_X = 0..4
  VALID_Y = 0..4
  VALID_F = [:NORTH,:SOUTH,:EAST,:WEST]
  TURN_RIGHT = {:SOUTH=>:WEST,:NORTH=>:EAST,:EAST=>:SOUTH,:WEST=>:NORTH}
  TURN_LEFT = {:SOUTH=>:EAST,:NORTH=>:WEST,:EAST=>:NORTH,:WEST=>:SOUTH}
  MOVE = {:SOUTH=>[0,-1],:NORTH=>[0,1],:EAST=>[1,0],:WEST=>[-1,0]}

  def initialize
    @accepting=false
  end

  def place(cols,rows,facing)
    x=cols.to_i
    y=rows.to_i
    f=facing.to_sym
    if is_valid?(x,y,f)
      @x=x
      @y=y
      @f=f
      @accepting=true
    end
  end

  def left
      return unless @accepting
      @f=TURN_LEFT[@f]
  end

  def right
      return unless @accepting
      @f=TURN_RIGHT[@f] 
  end

  def move
      return unless @accepting
      delta=MOVE[@f]
      x = @x + delta[0]
      y = @y + delta[1]
      if is_valid?(x,y)
        @x=x
        @y=y
      end
  end

  def report
    return nil unless @accepting
    "#{@x.to_s},#{@y.to_s},#{@f}"
  end

protected
  def is_valid?(x,y,f=@f)
      VALID_X.include?(x) && VALID_Y.include?(y) && VALID_F.include?(f)
  end 
end

File: gfvRobotJava/Robot/src/gfv/robot/Robot.java
package gfv.robot;

public class Robot {

	private Position currentPosition=null;


	/**
	 * MOVE will move the toy robot one unit forward 
	 * in the direction it is currently facing.
	 * 
	 */
	public void move() {
		if (currentPosition==null) return;
		Position newPosition = currentPosition.advance();
		place(newPosition);
	}
	/**
	 * REPORT will announce the X,Y and F of the robot. 
	 * This can be in any form, but standard output is sufficient.
	 * 
	 */
	public Position report() {
		if (currentPosition==null) return null;
		return currentPosition;
	}
	/**
	 * - LEFT and RIGHT will rotate the robot 90 degrees in 
	 * 	the specified direction without changing the position of the robot.
	 * 
	 * NOTE: I understand left as counterclockwise. So left when south is east
	 */
	public void left() {
		if (currentPosition==null) return;
		Position newPosition = currentPosition.turnLeft();
		place(newPosition);
	}
	/**
	 * - LEFT and RIGHT will rotate the robot 90 degrees in 
	 * 	the specified direction without changing the position of the robot.
	 * 
	 * 	NOTE: I understand right as clockwise. So right when south is west
	 */
	public void right() {
		if (currentPosition==null) return;
		Position newPosition = currentPosition.turnRight();
		place(newPosition);
		
	}
	/**
	 * - PLACE will put the toy robot on the table in position X,Y 
	 * and facing NORTH,  SOUTH, EAST or WEST.

	 * We want it to be thread safe, so position are never reused.
	 * @param strPosition String containing the Position 
	 */
	/*
	 */
	public synchronized void place(Position newPosition) {
		if (newPosition==null) return;
		if(!Board.isValidPosition(newPosition)) return;
		currentPosition=newPosition;
	}



}
